Detect non-upgraded Attic repositories

lfos opened the pull request:
When coming accross an invalid segment, explicitly check for the Attic
segment MAGIC. If an Attic segment is detected, raise a dedicated
exception, telling the user to upgrade the repository first.

Fixes #1933.

ThomasWaldmann commented on src/borg/repository.py:
```diff
@@ -872,7 +879,7 @@ def report_error(msg):
+            except (IntegrityError, LoggedIO.AtticSegment) as err:
                 report_error(str(err))
                 objects = []
                 if repair:
```
hmm, is it correct if it would attempt to repair here?

lfos replied:
If am not 100% sure. Looking at the code, it seems like a repair would convert the Attic segments into proper borg segments. So maybe, we would end up with a semi-converted repository? Or does `borg check` also take care of other places where things need to be converted and perform a full conversion? Either way, this is just backwards-compatible behavior. If this should be fixed, maybe in another commit -- see #2444?

ThomasWaldmann replied:
hmm...

i see 2 options:

- catch the AtticSegment exception separately and re-raise it as AtticRepository exception (== "check --repair" is not "upgrade from attic")
- actually change "check --repair" to be the "upgrade from attic" - there is a ticket open about the upgrade only being able to work locally, but if we would do it via "borg check" in the repository code, it could work remotely also.

to not have scope-creep in here, maybe do the first here and we can discuss doing the second later.

@enkore ^

lfos replied:
I just realized that we will actually never catch `AtticSegment` here. We already call `get_segments_transaction_id()` much earlier in the `check()` function. So if we decide to catch that error during `check`, it needs to be done much earlier. I tend to remove the extra check for `AtticSegment` here (leave this part of code untouched) and add proper Attic repository handling to `check` instead.

lfos commented:
Pushed a new version that addresses your remarks. I also fixed the base class of `AtticSegment` (should be `Error` instead of `Exception`). Thanks for the review!

ThomasWaldmann commented:
added to 1.1 backports. guess we really need this in 1.1, maybe even in 1.0.

in 1.2 it will be somehow pointless as release time of 1.2 is quite a bit in the future - maybe no attic repos left by then?

lfos commented:
Added Attic repository detection to `check` which is now also covered in the test case.

ThomasWaldmann commented:
Btw, when doing changes, rather add changeset than amending the existing one - easier to review and see what's changed. At the end, after review is finished, changesets can be collapsed to 1 or few.

lfos commented:
Good point. I am rather used to the ML-based patch submission process (as used for the Linux kernel or Git development itself) where it is more common to resubmit amended patch series. It is kind of annoying that there seems to be no easy way to display inter-diffs in GitHub... In the future, I will submit changes on top and squash only when the whole series is ready to be merged.

ThomasWaldmann commented on src/borg/repository.py:
```diff
@@ -321,7 +325,10 @@ def get_index_transaction_id(self):
+        try:
+            segments_transaction_id = self.io.get_segments_transaction_id()
+        except LoggedIO.AtticSegment:
+            raise self.AtticRepository(self.path)
```
more clear tracebacks using this:
```
except ... as e:
    raise ... from e
```

lfos replied:
Not sure about this one. The whole purpose of this patch is to get rid of the tracebacks and show a short and meaningful error message when running borg commands on Attic repositories...

ThomasWaldmann replied:
well, the point was to have nice traceback information.
whether we then show it or not is decided elsewhere.

ThomasWaldmann commented on src/borg/repository.py:
```diff
@@ -843,6 +850,8 @@ def report_error(msg):
             current_index = self.open_index(transaction_id)
             logger.debug('Read committed index of transaction %d', transaction_id)
+        except self.AtticRepository as exc:
+            raise
```
same here

lfos replied:
This is even more unclear to me. What do you want this to look like? Anyway, even if we decide to keep the simple `raise` here, the `as exc` part can go so this hunk will change in a fix-up commit.

ThomasWaldmann replied:
yes, just remove the "as exc". the plain "raise" re-raises the existing exception.

enkore replied:
What would be the point in re-raising? Since it's an `Error` the traceback is not normally shown, so there's little point in attaching a new traceback?

ThomasWaldmann replied:
the point of catching/reraising is so that is not catched by the `except Exception` below.

ThomasWaldmann commented on src/borg/repository.py:
```diff
@@ -1084,6 +1093,9 @@ class LoggedIO:
         """raised when a segment is full, before opening next"""
 
+    class AtticSegment(Error):
+        """Attic segment detected [segment {}]"""
```
'Attic segment magic detected [segment {}, offset {}]'.format(segment, 0)

lfos replied:
Huh? I am pretty sure you cannot do this in a doc string... Do you mean `"""Attic segment detected [segment {}, offset 0]"""` instead?

ThomasWaldmann replied:
ehrm, yes, the .format() happens elsewhere.

ThomasWaldmann replied:
the idea was just that it looks more consistent with the other magic related exception message.

lfos replied:
The Attic MAGIC will always be at offset 0, so I think it is fine to hardcode the value here.

ThomasWaldmann replied:
the other magic related message also specifically mentions offset 0.
this is to be consistent with other error messages always giving segment and offset.

ThomasWaldmann commented on src/borg/testsuite/repository.py:
```diff
@@ -714,6 +714,14 @@ def test_repair_missing_segment(self):
+            fd.seek(0)
+            fd.write(ATTIC_MAGIC)
+        self.assert_raises(Repository.AtticRepository, lambda: self.get_objects(4))
+        self.assert_raises(Repository.AtticRepository, lambda: self.check())
```
looks ok. another way would be to unpack the attic.tgz we have in the testsuite (which contains a repo made by attic).

ThomasWaldmann replied:
hmm, maybe copy the more highlevel test from my older try to fix this?

https://github.com/borgbackup/borg/pull/2207/files#diff-a74dadd5d28b08a8516d608b345d6f00R2269

lfos commented on src/borg/remote.py:
```diff
@@ -370,6 +370,8 @@ def inject_exception(self, kind):
         elif kind == 'IntegrityError':
             raise IntegrityError(s1)
+        elif kind == 'AtticRepository':
+            raise Repository.AtticRepository(s1)
```
I had the impression that we want to be able to test any exception with this function but that does not seem to be the case. Will remove this hunk.

lfos commented on src/borg/repository.py:
```diff
@@ -29,6 +29,7 @@
 
 MAGIC = b'BORG_SEG'
+ATTIC_MAGIC = b'ATTICSEG'
 MAGIC_LEN = len(MAGIC)
```
Looks good, will do.

ThomasWaldmann commented:
btw, another thing i noticed is that the attic magic is also defined in upgrader.py.

lfos commented:
True. We could import it from there but then we would lose the new assertion. Maybe both magic values should be defined in a central location (where we also have the length assertion)? Seems like this is going a bit out of the scope of this patch... But sounds like a nice idea for a follow-up patch!

ThomasWaldmann commented:
To avoid cyclic imports we maybe rather move it to repository.py. constants.py might be also an option.
Sure, can be in another patch.

ThomasWaldmann commented:
Hmm, btw, did you check that this handles all use cases when a user runs borg against a attic repo (no matter which command is used)?

lfos commented:
Yes, I checked that and no, we do not obtain an error message on every command that is run against an Attic repository. This only affects commands actually reading segments. Running a command like `borg break-lock` against an Attic repository works fine. Should we really error out there as well? If we do, we might as well scrap this patch (keep the `IntegrityError`) and add a check that is run every time before borg accesses a repository... What do you think?

ThomasWaldmann commented:
@enkore any feedback from you? would like to merge / backport this soon.

ThomasWaldmann commented:
@lfos no, i don't think we need to invest extra work on special stuff like `break-lock` or `debug ...`.

It's fine if the usual commands fail with a nice error message (e.g. create, extract, list, delete, prune).

enkore commented on src/borg/testsuite/repository.py:
```diff
@@ -714,6 +714,14 @@ def test_repair_missing_segment(self):
+        self.assert_raises(Repository.AtticRepository, lambda: self.get_objects(4))
+        self.assert_raises(Repository.AtticRepository, lambda: self.check())
+
     def test_repair_missing_commit_segment(self):
```
(No need to change it, just mentioning it) - the test suite was developed on unittest, but new code is allowed and encouraged to use pytest features like `with pytest.raises(SomeException): ...`

enkore commented:
LGTM

ThomasWaldmann commented:
@lfos still working on it or finished?

lfos commented:
I noticed some issues. Almost done with a new implementation. Will provide more details and an updated patch in a couple of minutes...

lfos commented:
I won't be able to finish the patch today. Have something that works more or less but I still need to rewrite the tests. The problem is that some operations (such as `borg create`) read segments only after making modifications which results in the Attic error message not being shown for some operations and being shown only after things have been modified for others. The new implementation always tries to read the latest segment (and check its MAGIC) when opening a repository, unless this check is explicitly disabled by an optional parameter (which is currently only used for `borg upgrade`).

lfos commented:
Pushed the rewrite (decided to rebase since the new commit almost has nothing in common with the original implementation). It also contains a "high-level test case" to make sure the usual commands are covered.

ThomasWaldmann commented on src/borg/testsuite/archiver.py:
```diff
@@ -2725,6 +2726,27 @@ def test_extract_hardlinks(self):
+        path = attic_repo(self.repository_path)
+        cmds = [
+            ['create', path + '::test', self.tmpdir],
+            ['extract', path + '::test', 'output.tar'],
```
output.tar?

lfos replied:
Good catch. Should be dropped. I will wait for further comments/reviews before submitting a fix, though...

lfos commented on src/borg/testsuite/upgrader.py:
```diff
@@ -149,8 +149,8 @@ def test_convert_all(attic_repo, attic_key_file, inplace):
-    # check should fail because of magic number
-    assert not repo_valid(repo_path)
+    with pytest.raises(Repository.AtticRepository):
+        repo_valid(repo_path)
```
Nowhere. It is defined in `src/borg/testsuite/upgrader.py` and does not have any other call sites.

ThomasWaldmann replied:
ok, i've looked it up now.
it is a bit of strange now as it can raise AtticRepository or return True/False for ok / not ok.
but i guess that is ok to deal with this special case, so keep it.

lfos commented:
Pushed an updated patch. Should be ready to be merged.

enkore commented:
Thanks!

ThomasWaldmann commented:
@lfos can you do the backport to 1.1-maint branch?

lfos commented:
Not sure what to do there, the commit applies fine against `1.1-maint`, so there isn't anything to do...

ThomasWaldmann commented:
Well, then it is just the PR.
